compiler/sem/semtypes.nim

```diff
@@ -232,12 +228,32 @@ proc semVarOutType(c: PContext, n: PNode, prev: PType; kind: TTypeKind): PType =
   else:
     result = newConstraint(c, kind)
 
+proc isRecursiveType(t: PType, cycleDetector: var IntSet): bool =
+  if t.isNil:
+    return false
+  if cycleDetector.containsOrIncl(t.id):
+    true
+  else:
+    case t.kind
+    of tyAlias, tyGenericInst, tyDistinct:
+      isRecursiveType(t.lastSon, cycleDetector)
+    else:
+      false
+
```

Comment:
Given how it's currently used, this can be simplified to something lile:
```suggestion
proc isReachable(t: PType, from: PType) =
  var curr = from
  while curr != t and curr.kind in {tyAlias, tyGenericInst, tyDistinct}:
    curr = curr.lastSon

  result = curr == t
```
This does require that self-recursive alias and distinct types weren't introduced earlier, however.

Reply:
'reachablility' is also what we're really testing here, and it's not always recursion (occurence of type application) because things like alias and forward are not instances of type application.

Reply:
Updated, thanks again for the insight.

Reply:
Naming it `aliasesType` like you did is probably the better choice.